### PR TITLE
update to PyO3 0.21.0-beta.0

### PR DESCRIPTION
**Pull request by davidhewitt**
This is a necessary precursor to trying the new PyO3 API in `pydantic-core`.

### CODE REVIEW

File: benches/python.rs [hunk 1]
```diff
@@ -10,25 +9,19 @@ use pyo3::Python;
...
 }
 
 fn python_parse_other(bench: &mut Bencher) {
     Python::with_gil(|py| {
-        bench.iter(|| {
-            // Clear PyO3 memory on each loop iteration to avoid long GC traversal overheads.
-            let _pool = unsafe { py.new_pool() };
-            black_box(python_parse(py, black_box(br#"["string", true, false, null]"#), false)).unwrap()
-        });
+        bench.iter(|| python_parse(py, br#"["string", true, false, null]"#, false).unwrap());
```
Comment (samuelcolvin) [inline benches/python.rs:24]: do we need a `black_box` around the return values?

**Review by samuelcolvin — COMMENTED** [review]
I'm interested that there isn't a bigger performance gain, also there seem to be lots of stacked new frames, is that expected?

image

File: benches/python.rs (hunk 1 above)
Comment (davidhewitt) [reply benches/python.rs:24]: I couldn't see any evidence that we needed it, and it doesn't seem to have affected our benchmarks.

**Comment by davidhewitt** [comment]
> I'm interested that there isn't a bigger performance gain, also there seem to be lots of stacked new frames, is that expected?

I did a bit of profiling and it looks like there's _a lot_ of Python overheads creating and destroying objects, so I don't think there's a huge amount more to be won in jiter. My gut tells me it'll matter more in `pydantic-core` where there's a lot more going on and fewer new objects being created / destroyed.

Re the stacked frames, the debug stack isn't a 1:1 mapping with the runtime call stack, a lot of stuff gets inlined and then expanded back out from debuginfo. So I'm not concerned by that.

**Comment by samuelcolvin** [comment]
Ye I was just surprised how much the flamegraph had changed, both in its shape and how many methods are marked as  "unknown" compared to before.

Could that be because we're using pyo3 from a git branch and something doesn't know how to find the names?

**Comment by davidhewitt** [comment]
Yes possibly, I don't have a good explanation as to why codspeed is failing to get those names.

**Pull request merged**